persistent_queue/persistent_queue.py

```diff
@@ -0,0 +1,464 @@
+            raise ValueError("filename must be set, e.g. '/my/path/important.queue'")
+        if maxsize < 0:
+            maxsize = 0
+
+        self.maxsize = maxsize
+        self.filename = os.path.abspath(filename)
+        self.dumps = dumps
+        self.loads = loads
+        self.flush_limit = flush_limit
+
+        self._file = self._open_file()
+        self._file_lock = threading.RLock()
+        self._get_lock = threading.RLock()
+        self._get_event = threading.Event()
+        self._put_lock = threading.RLock()
+        self._put_event = threading.Event()
+
+        self._all_tasks_done = threading.Condition()
+        self._unfinished_tasks = 0
+
+        self._file.seek(0, 0)
+        self._length = struct.unpack(HEADER_STRUCT[0], self._file.read(4))[0]
+
+    def _open_file(self, mode=None):
+        mode = mode or 'r+b' if os.path.isfile(self.filename) else 'w+b'
+        file = open(self.filename, mode=mode, buffering=0)
+
+        if mode == 'w+b':
+            # write length and start pointer
+            file.write(struct.pack(HEADER_STRUCT, 0, START_OFFSET))
+
+        return file
+
+    def _update_length(self, length):
+        current_pos = self._file.tell()
+
+        self._file.seek(0, 0)  # Go to the beginning of the file
+        self._file.write(struct.pack(HEADER_STRUCT[0], length))
+        self._file.flush()  # Probably not necessary since buffering=0
+        os.fsync(self._file.fileno())
+
+        self._length = length
+
+        self._file.seek(current_pos, 0)
+
+    def _get_queue_top(self):
+        current_pos = self._file.tell()
+
+        self._file.seek(START_OFFSET - 4, 0)  # Start at beginning of file
+        pos = struct.unpack(HEADER_STRUCT[1], self._file.read(4))[0]
+
+        self._file.seek(current_pos, 0)
+        return pos
+
+    def _set_queue_top(self, top):
+        current_pos = self._file.tell()
+
+        self._file.seek(START_OFFSET - 4, 0)  # Start at beginning of file
+        self._file.write(struct.pack(HEADER_STRUCT[1], top))
+        self._file.flush()  # Probably not necessary since buffering=0
+        os.fsync(self._file.fileno())
+
+        self._file.seek(current_pos, 0)
+
+    def _peek(self, block, timeout, items, partial=False):
+        """
+        Returns a certain amount of items from the queue. If items is greater
+        than one, a list is returned.
+        """
+        def read_data():
+            length = struct.unpack(LENGTH_STRUCT, self._file.read(4))[0]
+            data = self._file.read(length)
+            return self.loads(data)
+
+        _LOGGER.debug("Peeking %s items", items)
+
+        # Ignore requests for zero items
+        if items == 0:
+            _LOGGER.debug("Returning empty list")
+            return [], self._file.tell()
+
+        if block:
+            if timeout is not None:
+                target = time.time() + timeout
+            while self._length < items:
+                if self._put_event.wait(timeout) is not True:
```

Comment:
This part is within the `if block:` condition - so we should not raise a `queue.Empty` unless a timeout value is given.

If `timeout == None`, the code looks like this (simplified):
```python
            while self._length < items:
                self._put_event.wait(None) # only returns True
                self._put_event.clear()
```

If `timeout` is a number, the code looks like this (simplified)
`target` is going to be e.g. 5 seconds in the future.
```python
            while self._length < items:
                if self._put_event.wait(5) is not True:
                        timeout = target - time.time()
                        if timeout <= 0:  # only raise if timeout was set and has expired
                            raise queue.Empty
                # else: the event obviously returned True, which means a new item was pushed
                # so we just clear the event, and let the while-condition decode what to do next
                self._put_event.clear()
```